### PR TITLE
Search

### PR DESCRIPTION
**Pull request by mattheu**
live search. Uses the same code for live search as `hub.hmn.md`. I copy-pasted most of it. Falls back to normal WP_Query if the elastic search plugin is not activated.

### CODE REVIEW

File: assets/src/scripts/components/Search.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import 'whatwg-fetch';
+import debounce from 'lodash/debounce';
+import template from 'lodash/template';
+
+var init = function( searchBar ) {
+
+	var searchField      = searchBar.querySelector( '#site-search' ),
+	    resultsContainer = searchBar.querySelector( '#site-search-results' );
```
Comment (sambulance) [inline assets/src/scripts/components/Search.js]: As these are IDs, best to use `getElementById` here.

Comment (mattheu) [inline assets/src/scripts/components/Search.js]: The reason I used querySelector, is that it allows you to only select children of searchBar, but getElementById needs to be document wide.

I think i'm going to switch to a class. There's no reason you couldn't have multiple search fields on a page, and the JS implementation pretty much allows for that already, so lets not tie it to an ID.

Comment (sambulance) [inline assets/src/scripts/components/Search.js]: Ah I see. Class would be a better option 👍

File: inc/search.php [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+		$args['after'] = is_numeric( $args['after'] ) ? strtotime( $args['after'] . '-01-01' ) :   strtotime( $args['after'] );
+	}  else {
+		$args['after'] = null;
+	}
+
+	if ( ! empty( $args['user'] ) && ! is_numeric( $args['user'] )  ) {
+
+		global $wpdb;
+
+		$r = $wpdb->get_var( $wpdb->prepare( "SELECT ID FROM $wpdb->users WHERE user_login = %s OR display_name LIKE %s ", $args['user'], '%' . $args['user'] . '%' ) );
```
Comment (sambulance) [inline inc/search.php]: I'm not too familiar with Elastic search, but not sure why we are doing a database query here? Aren't we just preparing the query for search? Do we need to be able to search for users? Sorry if I'm misunderstanding

File: searchform.php [hunk 3]
```diff
@@ -1,11 +1,13 @@
 <div class="StyleGuide_SearchContainer">
 	<form class="SearchBar" role="search" method="get" action="<?php echo esc_url( home_url( '/' ) ); ?>">
 
-		<label class="SearchBar_Label" for="searchBar_Field"><?php esc_html_e( 'Search', 'hm-handbook' ); ?></label>
+		<label class="SearchBar_Label" for="SearchBar_Field"><?php esc_html_e( 'Search', 'hm-handbook' ); ?></label>
 
 		<div class="SearchBar_Container">
-			<input class="SearchBar_Field" type="search" id="searchBar_Field" placeholder="<?php esc_html_e( 'Search the site &hellip;', 'hm-handbook' ); ?>" value="<?php the_search_query(); ?>" name="s"/>
+			<input class="SearchBar_Field" type="search" id="site-search" placeholder="<?php esc_html_e( 'Search the site &hellip;', 'hm-handbook' ); ?>" value="<?php the_search_query(); ?>" name="s"/>
```
Comment (sambulance) [inline searchform.php]: This needs to match the label's `for`, so should be `SearchBar_Field`

**Comment by sambulance** [comment]
Given it a once over but maybe get another pair of eyes before merging :)

File: inc/search.php (hunk 2 above)
Comment (mattheu) [inline inc/search.php]: I'm not too familiar either, and I copy/pasted from the Hub site implementation. 

I think there are some fancy search thing supported such as `user:Tom` to fetch posts by a single user. Or `site:` or restricting by date.

Actually, whilst all of these things are nice and work particularly well on the Hub, i'm not sure how useful they are on the handbook. Maybe I just remove it all? If it stays, we need some way to show what filters are available.

Comment (mattheu) [inline inc/search.php]: Thought about this a little more... I don't think these filters are relevant to this site at all. Removed.

**Comment by mattheu** [comment]
@sambulance I've added a little more to this PR.

- Display a title on the real search results page for non-JS fallback.
- Add pagination to index pages just in case you ever end up hitting these.

**Comment by mattheu** [comment]
>  maybe get another pair of eyes before merging :)

If you think it looks OK i'm happy to merge! Just good to get someone to sanity check it a little :)

File: inc/search.php [hunk 4]
```diff
@@ -19,15 +19,15 @@
 
 add_action( 'rest_api_init', function() {
 
-	register_rest_route( 'hm-handbook/v1', '/search', array(
+	register_rest_route( 'hm-handbook/v1', '/search', [
```
Comment (sambulance) [inline inc/search.php]: Didn't know about short array syntax. Is it safe to use? Looks like it was introduced in PHP 5.4, so maybe avoid for public work?

Comment (mattheu) [inline inc/search.php]: If we're not releasing on the theme repo I don't think it matters. I'm happy to require PHP5.4+

Comment (mattheu) [inline inc/search.php]: Even 5.5 is past EOL now too :) http://php.net/supported-versions.php

**Pull request merged**